### PR TITLE
rockchip: add RK3399 NanoPi R4S 

### PR DESCRIPTION
**Pull request by tian-xiaobo**
Screen Shot 2020-12-03 at 2.26.12 PM|690x433 

* SoC – Rockchip RK3399 hexa-core processor with dual-Core Cortex-A72 up to 2.0GHz,  quad-core Cortex-A53 up to 1.5GHz, Mali-T864 GPU with OpenGL ES1.1/2.0/3.0/3.1, OpenCL, DX11, and AFBC support, 4K VP9 and 4K 10-bit H265/H264 60fps video decoder
* System Memory – 1GB DDR3 or 4GB LPDDR4
* Storage – MicroSD card slot
* Networking – 2x GbE(RTL8211E 1Gbps - RTL8111H 1Gbps), including one native Gigabit Ethernet, and one PCIe Gigabit Ethernet
* USB – 2x USB 3.0 Type-A ports, USB 2.0 via 4-pin header
* Expansion – 2×5-pin header with 1x SPI, 1x I2C
* Debugging – 3-pin debug UART header
* Misc- 1x power LED, and 3x user LEDs (SYS, LAN, WAN), user button, 2-pin RTC battery connector, 5V fan connector
* Power Supply
  * 5V/3A via USB-C connector or pin header
  * RK808-D PMIC and independent DC/DC enabling DVFS, software power-down, RTC wake-up, system sleep mode
* Dimensions – 66 x 66 mm (8-layer PCB)
* Temperature Range – -20°C to 70°C

The detailed information for [NanoPi R4S](https://wiki.friendlyarm.com/wiki/index.php/NanoPi_R4S)

Screen Shot 2020-12-06 at 5.04.02 PM|690x85

Signed-off-by: Xiaobo Tian tianxiaobo@gelaxy.io

### CODE REVIEW

File: package/boot/uboot-rockchip/Makefile [hunk 1]
```diff
@@ -5,10 +5,10 @@
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/kernel.mk
 
-PKG_VERSION:=2020.07
-PKG_RELEASE:=3
```
Comment (blocktrron) [inline package/boot/uboot-rockchip/Makefile:9]: This needs the updated generated platform-data for the SPL, which is missing in this PR.

File: package/boot/uboot-rockchip/Makefile [hunk 2]
```diff
@@ -38,6 +38,26 @@ endef
...
+  BUILD_SUBTARGET:=armv8
+  NAME:=NanoPi R4S
+  BUILD_DEVICES:= \
+    friendlyarm_nanopi-r4s
+  DEPENDS:=+PACKAGE_u-boot-nanopi-r4s-rk3399:arm-trusted-firmware-rockchip
+  PKG_BUILD_DEPENDS:=arm-trusted-firmware-rockchip
+  ATF:=rk3399_bl31.elf
+endef
+
+define U-Boot/nanopi-r4s-1g-rk3399
```
Comment (blocktrron) [inline package/boot/uboot-rockchip/Makefile:51]: Consistent naming, as both models need a separate loader.

File: package/boot/uboot-rockchip/patches/200-uboot-rockchip-add-support-for-NanoPi-R4S.patch [hunk 3]
```diff
@@ -0,0 +1,227 @@
+From 6b2ff6193a7f402e70a7def00857b3c12e0b7891 Mon Sep 17 00:00:00 2001
+From: Marty Jones <mj8263788@gmail.com>
+Date: Sat, 12 Dec 2020 02:08:13 -0500
+Subject: [PATCH] uboot-rockchip: add support for NanoPi R4S
+
+Signed-off-by: Marty Jones <mj8263788@gmail.com>
```
Comment (blocktrron) [inline package/boot/uboot-rockchip/patches/200-uboot-rockchip-add-support-for-NanoPi-R4S.patch:6]: Please send this upstream

File: package/boot/uboot-rockchip/patches/201-uboot-rockchip-add-NanoPi-R4S-1Gb-DDR3.patch [hunk 4]
```diff
@@ -0,0 +1,105 @@
+From 2d51eeb1dd80b5fa030cd2b2962eeff7599af9a3 Mon Sep 17 00:00:00 2001
+From: Marty Jones <mj8263788@gmail.com>
+Date: Sat, 12 Dec 2020 03:23:32 -0500
+Subject: [PATCH] uboot-rockchip: add NanoPi R4S 1Gb DDR3
```
Comment (blocktrron) [inline package/boot/uboot-rockchip/patches/201-uboot-rockchip-add-NanoPi-R4S-1Gb-DDR3.patch:4]: Same for this.

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity [hunk 5]
```diff
@@ -26,5 +26,13 @@ friendlyarm,nanopi-r2s)
...
 	set_interface_core 4 "eth1" "xhci-hcd:usb3"
 	;;
+friendlyarm,nanopi-r4s)
+	set_interface_core 2 "eth0"
+	set_interface_core 4 "eth1"
+	;;
+friendlyarm,nanopi-r4s-1g)
+	set_interface_core 2 "eth0"
+	set_interface_core 4 "eth1"
+	;;
```
Comment (blocktrron) [inline target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:36]: Any reason for different switch cases?

File: target/linux/rockchip/image/armv8.mk [hunk 6]
```diff
@@ -15,6 +15,26 @@ define Device/friendlyarm_nanopi-r2s
...
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPi R4S
+  SOC := rk3399
+  UBOOT_DEVICE_NAME := nanopi-r4s-rk3399
+  IMAGE/sysupgrade.img.gz := boot-common | boot-script nanopi-r4s | pine64-img | gzip | append-metadata
+  DEVICE_PACKAGES := kmod-r8169
+endef
+TARGET_DEVICES += friendlyarm_nanopi-r4s
+
+define Device/friendlyarm_nanopi-r4s-1g
```
Comment (blocktrron) [inline target/linux/rockchip/image/armv8.mk:28]: same here - more self-explanatory naming.

File: target/linux/rockchip/patches-5.4/110-rockchip-add-NanoPi-R4S-DTS.patch [hunk 7]
```diff
@@ -0,0 +1,162 @@
+From ca74fb56aaa3fe565bfe08574dcb0fd094736e28 Mon Sep 17 00:00:00 2001
+From: Marty Jones <mj8263788@gmail.com>
+Date: Sat, 12 Dec 2020 02:48:47 -0500
+Subject: [PATCH] rockchip: add NanoPi R4S DTS
+
```
Comment (blocktrron) [inline target/linux/rockchip/patches-5.4/110-rockchip-add-NanoPi-R4S-DTS.patch:5]: Description missing? Also send this upstream please.

File: target/linux/rockchip/patches-5.4/111-rockchip-add-NanoPi-R4S-1Gb-DDR3.patch [hunk 8]
```diff
@@ -0,0 +1,27 @@
+From dbbcf8b2cb8e321310510179caa12afb30525c17 Mon Sep 17 00:00:00 2001
+From: Marty Jones <mj8263788@gmail.com>
+Date: Sat, 12 Dec 2020 03:54:03 -0500
+Subject: [PATCH] rockchip: add NanoPi R4S 1Gb DDR3
+
```
Comment (blocktrron) [inline target/linux/rockchip/patches-5.4/111-rockchip-add-NanoPi-R4S-1Gb-DDR3.patch:5]: I don't get it.

If only the bootloader needs to be model-specific in order to initialize the DDR controller, what benefit does a specific kernel image provide at all?

File: target/linux/rockchip/patches-5.4/110-rockchip-add-NanoPi-R4S-DTS.patch [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
++	vpcie3v3-supply = <&vcc3v3_sys>;
++
++	pcie@0 {
++		reg = <0x00000000 0 0 0 0>;
++		#address-cells = <3>;
++		#size-cells = <2>;
++
++		r8169: pcie@0,0 {
++			reg = <0x000000 0 0 0 0>;
++			local-mac-address = [ 00 00 00 00 00 00 ];
```
Comment (blocktrron) [inline target/linux/rockchip/patches-5.4/110-rockchip-add-NanoPi-R4S-DTS.patch:131]: Any specific reason for setting the address to all-zero here?

File: target/linux/rockchip/patches-5.4/110-rockchip-add-NanoPi-R4S-DTS.patch [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
++
++/dts-v1/;
++#include "rk3399-nanopi4.dtsi"
++
++/ {
++	model = "FriendlyElec NanoPi R4S";
++	compatible = "friendlyarm,nanopi-r4s", "rockchip,rk3399";
++
++	aliases {
++		ethernet1 = &r8169;
```
Comment (blocktrron) [inline target/linux/rockchip/patches-5.4/110-rockchip-add-NanoPi-R4S-DTS.patch:44]: Add a second patch on the one upstreamed for the OpenWrt specific status LED assignment

**Comment by peterwillcn** [comment]
u-boot
https://github.com/u-boot/u-boot/pull/43
kernel
WIP

**Comment by blocktrron** [comment]
denx only accepts patches on the mailinglist, see https://github.com/u-boot/u-boot/blob/master/.github/pull_request_template.md

**Comment by peterwillcn** [comment]
@blocktrron Merged https://github.com/u-boot/u-boot/commit/b69b9f3f54732c303939eb748aad97cd4cf60168

**Pull request closed without merging**